src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs

```diff
@@ -77,503 +59,211 @@ public void Abort()
+                ValidateHeader(response, HttpKnownHeaderNames.Upgrade, "websocket");
+                ValidateHeader(response, HttpKnownHeaderNames.SecWebSocketAccept, secKeyAndSecWebSocketAccept.Value);
 
-                // Add all of the additionally requested headers
-                foreach (string key in options.RequestHeaders.AllKeys)
+                // The SecWebSocketProtocol header is optional.  We should only get it with a non-empty value if we requested subprotocols,
+                // and then it must only be one of the ones we requested.  If we got a subprotocol other than one we requested (or if we
+                // already got one in a previous header), fail. Otherwise, track which one we got.
+                string subprotocol = null;
+                IEnumerable<string> subprotocolEnumerableValues;
+                if (response.Headers.TryGetValues(HttpKnownHeaderNames.SecWebSocketProtocol, out subprotocolEnumerableValues))
                 {
-                    if (string.Equals(key, HttpKnownHeaderNames.Host, StringComparison.OrdinalIgnoreCase))
+                    Debug.Assert(subprotocolEnumerableValues is string[]);
+                    string[] subprotocolArray = (string[])subprotocolEnumerableValues;
+                    if (subprotocolArray.Length > 0 && !string.IsNullOrEmpty(subprotocolArray[0]))
                     {
-                        // Host header handled above
-                        continue;
+                        subprotocol = options.RequestedSubProtocols.Find(requested => string.Equals(requested, subprotocolArray[0], StringComparison.OrdinalIgnoreCase));
+                        if (subprotocol == null)
+                        {
+                            throw new WebSocketException(
+                                WebSocketError.UnsupportedProtocol,
+                                SR.Format(SR.net_WebSockets_AcceptUnsupportedProtocol, string.Join(", ", options.RequestedSubProtocols), string.Join(", ", subprotocolArray)));
+                        }
                     }
-
-                    builder.Append(key).Append(": ").Append(options.RequestHeaders[key]).Append("\r\n");
                 }
 
-                // Add the optional subprotocols header
-                if (options.RequestedSubProtocols.Count > 0)
-                {
-                    builder.Append(HttpKnownHeaderNames.SecWebSocketProtocol).Append(": ");
-                    builder.Append(options.RequestedSubProtocols[0]);
-                    for (int i = 1; i < options.RequestedSubProtocols.Count; i++)
+                // Get the response stream and wrap it in a web socket.
+                Stream connectedStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
+                Debug.Assert(connectedStream.CanWrite);
+                Debug.Assert(connectedStream.CanRead);
+                MethodInfo createFromStreamMethod = typeof(WebSocket).GetMethod(
```

Comment:
>WebSocket [](start = 59, length = 9)

Add a similar comment about how this is supported in netcoreapp2.1 but is not part of netstandard2.0.